### PR TITLE
Add DSNT operations to contrib

### PR DESCRIPTION
**Pull request by anibali**
Add the important bits of https://github.com/anibali/dsntnn as a contrib module.

Related issues: https://github.com/arraiyopensource/kornia/issues/138, https://github.com/arraiyopensource/kornia/issues/166

### CODE REVIEW

File: kornia/contrib/spatial_soft_argmax2d.py [hunk 1]
```diff
@@ -36,33 +35,9 @@ def spatial_soft_argmax2d(
...
-    pos_y: torch.Tensor = grid[..., 1].reshape(-1)
-
-    # compute the expected coordinates
-    expected_y: torch.Tensor = torch.sum(pos_y * x_soft, dim=-1, keepdim=True)
-    expected_x: torch.Tensor = torch.sum(pos_x * x_soft, dim=-1, keepdim=True)
-
-    output: torch.Tensor = torch.cat([expected_x, expected_y], dim=-1)
-    return output.view(batch_size, channels, 2)  # BxNx2
+    input_soft: torch.Tensor = dsnt.spatial_softmax_2d(input, temperature)
+    output: torch.Tensor = dsnt.spatial_softargmax_2d(input, normalized_coordinates)
```
Comment (willhnguyen) [inline kornia/contrib/spatial_soft_argmax2d.py]:
```suggestion
    output: torch.Tensor = dsnt.spatial_softargmax_2d(input_soft, normalized_coordinates)
```

Comment (anibali) [reply kornia/contrib/spatial_soft_argmax2d.py]: Good spot, fixed.

**Comment by anibali** [comment]
It looks like mypy is failing to detect two PyTorch attributes/methods: torch.dtype.is_floating_point (documented at https://pytorch.org/docs/stable/tensor_attributes.html#torch.torch.dtype) and torch.finfo (documented at https://pytorch.org/docs/stable/type_info.html#torch.torch.finfo). I don't normally use mypy, so I'm not sure how to work around this.

File: kornia/contrib/dsnt.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+    k_y = -0.5 * torch.reciprocal(std[..., 1, None, None])
+
+    # Assemble the 2D Gaussian.
+    exps_x = torch.exp(dist_x * k_x)
+    exps_y = torch.exp(dist_y * k_y)
+    gauss = exps_x * exps_y
+
+    # Rescale so that values sum to one.
+    val_sum = gauss.sum(-2, keepdim=True).sum(-1, keepdim=True)
+    val_sum = val_sum + torch.finfo(dtype).tiny  # Avoid division by zero.
```
Comment (edgarriba) [inline kornia/contrib/dsnt.py]: it could be useful to abstract this to an external function, I think something like that might work
```python
def safe_zero_division(numerator: torch.Tensor, denominator: torch.Tensor) -> torch.Tensor:
    eps: float = torch.finfo(numerator.dtype).tiny
    return numerator / torch.clamp(denominator, min=eps)
```
/cc @bhack what do you think?

File: kornia/contrib/dsnt.py [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+    gauss = gauss / val_sum
+
+    return gauss
+
+
+def _kl_div(p, q, ndims):
+    eps = 1e-24
+    unsummed_kl = p * ((p + eps).log() - (q + eps).log())
```
Comment (edgarriba) [inline kornia/contrib/dsnt.py]: can't you just use https://pytorch.org/docs/stable/nn.html?highlight=kl#torch.nn.functional.kl_div ?

File: kornia/contrib/dsnt.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+          summed. Default: ``'mean'``.
+
+    Shape:
+        - Input: :math:`(B, N, H, W)`
+        - Target: :math:`(B, N, H, W)`, same shape as the input
+    """
+    return _reduce_loss(_js_div(input, target, ndims=2), reduction)
+
+
+def kl_div_loss_2d(
```
Comment (edgarriba) [inline kornia/contrib/dsnt.py]: I'd put this under kornia.losses namespace

File: test/test_contrib.py [hunk 5]
```diff
@@ -239,3 +240,59 @@ def test_gradcheck(self):
...
+    def test_render_gaussian_2d_normalized_coordinates(self):
+        expected = self.GAUSSIAN
+        actual = dsnt.render_gaussian_2d(torch.tensor([0.0, 0.0]),
+                                         torch.tensor([0.25, 0.25]),
+                                         (5, 5),
+                                         normalized_coordinates=True)
+        assert_allclose(actual, expected, rtol=0, atol=1e-5)
+
+    def test_js_div_loss_2d(self):
+        examples = [
```
Comment (edgarriba) [inline test/test_contrib.py]: I'd use for this `@pytest.mark.parametrize` to generate separate test
Check: https://github.com/arraiyopensource/kornia/blob/master/test/geometry/test_conversions.py#L23

File: test/test_contrib.py [hunk 6]
```diff
@@ -239,3 +240,59 @@ def test_gradcheck(self):
...
+            hm = dsnt.spatial_softmax_2d(input, temperature)
+            pred = dsnt.spatial_softargmax_2d(hm, normalized_coordinates)
+            target_hm = dsnt.render_gaussian_2d(target, std, input.shape[-2:],
+                                                normalized_coordinates)
+            loss1 = mse_loss(pred, target, size_average=None, reduce=None,
+                             reduction='mean')
+            loss2 = dsnt.js_div_loss_2d(hm, target_hm, 'mean')
+            return loss1 + loss2
+
+        op_script = torch.jit.script(op)
```
Comment (edgarriba) [inline test/test_contrib.py]: that rocks ! Did you know that you can even serialize/de-serialize this op ? :D

File: kornia/contrib/dsnt.py [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+          Default is True.
+
+    Shape:
+        - `mean`: :math:`(*, 2)`
+        - `std`: :math:`(*, 2)`
+        - Output: :math:`(*, H, W)`
+    """
+    device = mean.device
+    dtype = mean.dtype
+    if not dtype.is_floating_point:
```
Comment (edgarriba) [inline kornia/contrib/dsnt.py]: for mypy if I don't have any clue about this, I don't really know why is not recognised at all. If that blocks you, use ` # type: ignore` to avoid mypy check in this line

Comment (anibali) [reply kornia/contrib/dsnt.py]: Using `# type: ignore` doesn't work in that case, as mentioned in this note from the mypy docs:

> The # type: ignore comment will only assign the implicit Any type if mypy cannot find information about that particular module. So, if we did have a stub available for frobnicate then mypy would ignore the # type: ignore comment and typecheck the stub as usual.

However, I can annotate the type of `dtype` as Any and that does work.

File: kornia/contrib/dsnt.py [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+        raise TypeError("Input type is not a torch.Tensor. Got {}"
+                        .format(type(tensor)))
+    if not len(tensor.shape) == 4:
+        raise ValueError("Invalid input shape, we expect BxCxHxW. Got: {}"
+                         .format(tensor.shape))
+
+
+def spatial_softmax_2d(
+        input: torch.Tensor,
+        temperature: torch.Tensor = torch.tensor(1.0),
```
[the remark this answers is not in the document]
Comment (anibali) [reply kornia/contrib/dsnt.py]: `typing.Optional` means that the value can be `None`. We don't support passing in `None`, hence the type should not be `Optional[torch.Tensor]`.

**Comment by anibali** [comment]
I added the code review fixes as a separate commit for the sake of clarity. The commits can be squashed later.

How about calling the new module `kornia.geometry.locations` or `kornia.geometry.localization`, since everything is focussed around finding locations of points of interest?

File: kornia/contrib/dsnt.py (hunk 2 above)
Comment (anibali) [reply kornia/contrib/dsnt.py]: I've broken it out into a local helper function. However, if you decide to make it a more accessible utility then it's important to make it crystal clear that it will only work properly for non-negative denominators.

File: kornia/contrib/dsnt.py [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+    Shape:
+        - `mean`: :math:`(*, 2)`
+        - `std`: :math:`(*, 2)`
+        - Output: :math:`(*, H, W)`
+    """
+    device = mean.device
+    dtype = mean.dtype
+    if not dtype.is_floating_point:
+        raise TypeError("Expected `mean` to have floating point values. Got {}"
+                        .format(dtype))
```
[the remark this answers is not in the document]
Comment (anibali) [reply kornia/contrib/dsnt.py]: Added for dtype and device, but shapes can actually be different. It's often convenient to just specify one standard deviation for everything. I clarified in the docstring that the requirement is that std and mean are broadcastable.

**Comment by edgarriba** [comment]
@anibali I'm not quite sure yet about the module name. Libraries like opencv or skimage they place all this stuff (or similar) inside `feature` like modules.

**Comment by anibali** [comment]
Well that name would depend on whether you consider the primary purpose of this code to be feature detection. I would argue that the main purpose of this code is to locate points of interest, and that a name like `feature` could make it harder for people to find.

**Comment by anibali** [comment]
Is there any intention to ever merge this? The implementation I wrote here is cleaner than my own library, so if you're not planning on using it I might just roll it back into dsntnn as a version 2.0 or something.

**Comment by edgarriba** [comment]
@anibali sorry we have been busy with the paper writing and taking some time to discuss next goals. If there's any objection let's keep this inside contrib by now. The code looks good to me. Let's wait for @ducha-aiki approval.

**Comment by edgarriba** [comment]
@anibali do you have more contributions in mind ? We are listing things to do to be included in the roadmap.

**Comment by edgarriba** [comment]
@ducha-aiki talk now or never 😂

**Comment by anibali** [comment]
@edgarriba I don't have a problem with code from this PR living in contrib for now. I can't think of any ideas for future contributions off the top of my head, but I'll let you know if something comes up. 

@ducha-aiki We have a few projects in our research group that are using soft-argmax/DSNT to locate points. If you know of improvements/better alternatives to the algorithm, please share your thoughts :)

**Comment by edgarriba** [comment]
@anibali you can disable jit tests, since the noticed that the current approach seems a bit unstable. We need to rethink about that.

**Comment by ducha-aiki** [comment]
@anibali I cannot find a reference now -- it was some 2017-2018 ML paper about planning/value iteration, where they showed that for highly symmetrical things, softargmax will provide very wrong answer. Like for array [ 0.98, 0 , 0, 0, 0, 0, 0, 0, 0.99], the output will be in the middle, instead of mode. So they proposed to take top-K and then do weighted random for training and just max for the test. I also have used this trick.
But, if course, it is domain dependent

@edgarriba softargmax2d it is nice to have algorithm in kornia for sure :)

**Comment by bhack** [comment]
I think that also could have issue when It Is spread. But in that case was proposed another op/loss https://arxiv.org/abs/1903.11633

**Comment by edgarriba** [comment]
@anibali extra thing I see, provide a test for this: https://codecov.io/gh/arraiyopensource/kornia/pull/167/diff?src=pr&el=tree#D1-28

**Comment by anibali** [comment]
@edgarriba I have increased test coverage and disabled the JIT test.

**Comment by edgarriba** [comment]
Aawesome. Merging
